src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+            private double[] _backAddArray;
+            private double[] _fftRe;
+            private double[] _fftIm;
+            private double[] _magList;
+            private double[] _magLogList;
+            private double[] _spectralList;
+            private double[] _transRe;
+            private double[] _transIm;
+            private double[] _ifftRe;
+            private double[] _ifftIm;
+            private double[] _ifftMagList;
+            private double[] _cumSumList;
+            private double[] _cumSumShift;
+            //used in AnomalyAndMargin mode
+            private double[] _units;
+            private double[] _val;
+            private double[] _ans;
+            private double[] _curWindow;
+
+            public SrCnnEntireModeler(double threshold, double sensitivity, SrCnnDetectMode detectMode)
+            {
+                _threshold = threshold;
+                _sensitivity = sensitivity;
+                _detectMode = detectMode;
+            }
+
+            public void Train(double[] values, ref double[][] results)
+            {
+                if (results == null)
+                {
+                    results = new double[values.Length][];
+                    for (int i = 0; i < results.Length; ++i)
+                    {
+                        results[i] = new double[_outputLengthArray[(int)_detectMode]];
+                    }
+                }
+                else if (results.Length > values.Length)
+                {
+                    Array.Resize<double[]>(ref results, values.Length);
+                }
+                SpectralResidual(values, results, _threshold);
+                //Optional Steps
+                if (_detectMode == SrCnnDetectMode.AnomalyAndMargin)
+                {
+                    GetMargin(values, results, _sensitivity);
+                }
+                else if (_detectMode == SrCnnDetectMode.AnomalyAndExpectedValue)
+                {
+                    GetExpectedValue(values, results);
+                }
+            }
+
+            private void AllocateDoubleArray(ref double[] arr, int length)
+            {
+                if (arr == null)
+                {
+                    arr = new double[length];
+                }
+                else if (arr.Length != length)
+                {
+                    Array.Resize<double>(ref arr, length);
+                }
+            }
+
+            private void SpectralResidual(double[] values, double[][] results, double threshold)
+            {
+                // Step 1: Get backadd wave
+                double[] backAddList = BackAdd(values);
+
+                // Step 2: FFT transformation
+                int length = backAddList.Length;
+                AllocateDoubleArray(ref _fftRe, length);
+                AllocateDoubleArray(ref _fftIm, length);
+
+                FftUtils.ComputeForwardFft(backAddList, Enumerable.Repeat((double)0.0f, length).ToArray(), _fftRe, _fftIm, length);
```

Comment:
>ComputeForwardFft [](start = 25, length = 17)

I think you might be able to change this method to take in `Span<double>` instead of arrays, this way you won't have to resize, you can just take a subspan of the arrays, and the arrays will always be the same length.

Reply:
Explained in email

---
In reply to: [427857936](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427857936) [](ancestors = 427857936)